File: AndroidPIH/app/src/main/java/mx/cfe/grijalva/pih/data/service/StationDataService.kt
package mx.cfe.grijalva.pih.data.service

import android.content.Context
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import mx.cfe.grijalva.pih.data.model.*
import mx.cfe.grijalva.pih.data.network.NetworkModule
import java.text.SimpleDateFormat
import java.util.*

class StationDataService(private val context: Context, private val authService: AuthService) {
    private val TAG = "StationDataService"

    private val _stations = MutableStateFlow<List<StationInfo>>(emptyList())
    val stations: StateFlow<List<StationInfo>> = _stations

    private val _variables = MutableStateFlow<List<StationVariable>>(emptyList())
    val variables: StateFlow<List<StationVariable>> = _variables

    private val _analysisData = MutableStateFlow<DataAnalysisResponse?>(null)
    val analysisData: StateFlow<DataAnalysisResponse?> = _analysisData

    private val _isLoadingStations = MutableStateFlow(false)
    val isLoadingStations: StateFlow<Boolean> = _isLoadingStations

    private val _isLoadingVariables = MutableStateFlow(false)
    val isLoadingVariables: StateFlow<Boolean> = _isLoadingVariables

    private val _isLoadingData = MutableStateFlow(false)
    val isLoadingData: StateFlow<Boolean> = _isLoadingData

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    suspend fun loadStations(onlyCfe: Boolean = true) {
        _isLoadingStations.value = true
        _isLoading.value = true
        try {
            val api = NetworkModule.getApi(context)
            val response = api.getStations(authService.authHeader, onlyCfe)
            if (response.isSuccessful) {
                _stations.value = response.body() ?: emptyList()
            }
        } catch (e: Exception) {
            Log.e(TAG, "loadStations error", e)
        } finally {
            _isLoadingStations.value = false
            _isLoading.value = false
        }
    }

    suspend fun loadVariables(stationId: String) {
        _isLoadingVariables.value = true
        try {
            val api = NetworkModule.getApi(context)
            val response = api.getStationVariables(authService.authHeader, stationId)
            if (response.isSuccessful) {
                _variables.value = response.body() ?: emptyList()
            }
        } catch (e: Exception) {
            Log.e(TAG, "loadVariables error", e)
        } finally {
            _isLoadingVariables.value = false
        }
    }

    suspend fun loadAnalysisData(
        stationIds: List<String>, variable: String,
        startDate: Date, endDate: Date
    ) {
        val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        loadAnalysisData(stationIds, variable, fmt.format(startDate), fmt.format(endDate))
    }

    suspend fun loadAnalysisData(
        stationIds: List<String>, variable: String,
        startDate: String, endDate: String
    ) {
        _isLoadingData.value = true
        _isLoading.value = true
        _errorMessage.value = null
        try {
            val api = NetworkModule.getApi(context)
            val response = api.getAnalysisData(
                authService.authHeader,
                AnalysisRequest(stationIds, variable, startDate, endDate)
            )
            if (response.isSuccessful) {
                _analysisData.value = response.body()
            } else {
                _errorMessage.value = "Error ${response.code()}"
            }
        } catch (e: Exception) {
            _errorMessage.value = "Error: ${e.localizedMessage}"
            Log.e(TAG, "loadAnalysisData error", e)
        } finally {
            _isLoadingData.value = false
            _isLoading.value = false
        }
    }
}

File: ChatiOS/CloudStationChat/Services/StationDataService.swift
import Foundation
import Combine

@MainActor
class StationDataService: NSObject, ObservableObject, URLSessionDelegate {
    @Published var stations: [StationInfo] = []
    @Published var variables: [StationVariable] = []
    @Published var analysisData: DataAnalysisResponse?
    @Published var isLoadingStations = false
    @Published var isLoadingVariables = false
    @Published var isLoadingData = false
    @Published var errorMessage: String?
    
    private var serverUrl: String = ""
    private var token: String = ""
    private var urlSession: URLSession!
    
    override init() {
        super.init()
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        config.httpAdditionalHeaders = ["Accept": "application/json"]
        config.protocolClasses = nil
        urlSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
    }
    
    func configure(serverUrl: String, token: String) {
        self.serverUrl = serverUrl
        self.token = token
    }
    
    // MARK: - TLS bypass
    nonisolated func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                                completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        if let trust = challenge.protectionSpace.serverTrust {
            completionHandler(.useCredential, URLCredential(trust: trust))
        } else {
            completionHandler(.performDefaultHandling, nil)
        }
    }
    
    // MARK: - Load Stations
    func loadStations(onlyCfe: Bool = true) async {
        guard !serverUrl.isEmpty else { return }
        isLoadingStations = true
        errorMessage = nil
        defer { isLoadingStations = false }
        
        do {
            let url = URL(string: "\(serverUrl)/DataAnalysis/GetStations?onlyCfe=\(onlyCfe)")!
            var request = URLRequest(url: url)
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            
            let (data, response) = try await urlSession.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                errorMessage = "Error cargando estaciones"
                return
            }
            stations = try JSONDecoder().decode([StationInfo].self, from: data)
        } catch {
            errorMessage = "Error: \(error.localizedDescription)"
        }
    }
    
    // MARK: - Load Variables for Station
    func loadVariables(stationId: String) async {
        guard !serverUrl.isEmpty else { return }
        isLoadingVariables = true
        variables = []
        defer { isLoadingVariables = false }
        
        do {
            let encoded = stationId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? stationId
            let url = URL(string: "\(serverUrl)/DataAnalysis/GetStationVariables?stationId=\(encoded)")!
            var request = URLRequest(url: url)
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            
            let (data, response) = try await urlSession.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                return
            }
            variables = try JSONDecoder().decode([StationVariable].self, from: data)
        } catch {
            print("Error loading variables: \(error)")
        }
    }
    
    // MARK: - Load Analysis Data
    func loadAnalysisData(stationIds: [String], variable: String, startDate: Date, endDate: Date) async {
        guard !serverUrl.isEmpty else { return }
        isLoadingData = true
        errorMessage = nil
        analysisData = nil
        defer { isLoadingData = false }
        
        do {
            let url = URL(string: "\(serverUrl)/DataAnalysis/GetAnalysisData")!
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            
            let iso = ISO8601DateFormatter()
            let body = DataAnalysisRequest(
                stationIds: stationIds,
                variable: variable,
                startDate: iso.string(from: startDate),
                endDate: iso.string(from: endDate)
            )
            request.httpBody = try JSONEncoder().encode(body)
            
            let (data, response) = try await urlSession.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                errorMessage = "Error del servidor"
                return
            }
            analysisData = try JSONDecoder().decode(DataAnalysisResponse.self, from: data)
        } catch {
            errorMessage = "Error: \(error.localizedDescription)"
        }
    }
}
